### ISSUE TITLE
Bug: [unbound-method] should flag on method in object literal defined using longhand form

### ISSUE THREAD

**Issue by kirkwaiblinger** [typescript-eslint/typescript-eslint#8636]
### Before You File a Bug Report Please Confirm You Have Done The Following...

- [X] I have tried restarting my IDE and the issue persists.
- [X] I have updated to the latest version of the packages.
- [X] I have [searched for related issues](https://github.com/typescript-eslint/typescript-eslint/issues?q=is%3Aissue+label%3A%22package%3A+eslint-plugin%22) and found none that matched my issue.
- [X] I have [read the FAQ](https://typescript-eslint.io/linting/troubleshooting) and my problem is not listed.

### Playground Link

https://typescript-eslint.io/play/#ts=5.3.3&showAST=es&fileType=.tsx&code=MYewdgzgLgBCBGArApsKAZAllZAnAhgDYwC8MA3gFAwwBmAXHQK5hqbgwAUAlBdTTFzIoTXGBhQAFpggA6AOY8A3PwC%2BAGn6LeVATFCQQhZLMIhFAIgASyQmZgB3ELkIATAIQXuayqsqUDaDpSOCRUDGw8IllaFVplGAB6RJgAVVZ8JnlJWAAVAE8AB2QAUVxcZ0YpGQUYGRgwEFh8ZlYodjBKIA&eslintrc=N4KABGBEBOCuA2BTAzpAXGUEKQAIBcBPABxQGNoBLY-AWhXkoDt8B6WJgIwHsOATWgFtE%2BABbc%2B6KImjRu0SODABfEMqA&tsconfig=N4KABGBEDGD2C2AHAlgGwKYCcDyiAuysAdgM6QBcYoEEkJemy0eFYDAruuGAL4g9A&tokens=false

### Repro Code

```TypeScript
const objectLiteral = {
  f: function () {
    return this.g();
  },
  g() {
    console.log("Hello world!")
  }
}

const f = objectLiteral.f;
f(); // Uncaught TypeError: this.g is not a function
```

### ESLint Config

```javascript
module.exports = {
  "rules": {
    "@typescript-eslint/unbound-method": "error"
  }
}
```

### tsconfig

```jsonc
{
  "compilerOptions": {
    "strict": true
  }
}
```

### Expected Result

I expect that the rule should flag on the line `const f = objectLiteral.f;`

### Actual Result

No error

### Additional Info

The rule has different behavior depending whether a function is assigned to a property as a method in shorthand form (`f() {}`), a method in longhand form (`f: function () {}`), or an arrow function (`f: () => {}`) (see also [mdn link on object expressions](https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Operators/Object_initializer#method_definitions)). Expressions that coincidentally evaluate to a function and arrow function methods both error early due to TS complaining about the use of `this` when they are defined. And the _shorthand_ method form is not handled by TS, but it is correctly flagged by the rule. It's just the _longhand_ form that falls through the cracks.

The following snippet may be illustrative ([playground link](https://typescript-eslint.io/play/#ts=5.3.3&showAST=es&fileType=.tsx&code=MYewdgzgLgBCBGArApsKAZAllZAnAhgDYDKAFsmPmJgOZUQwC8MA3gFAwwC2yUpIAEzIhcfKkICeYKPgAeACgCUrDpxi5eAV1xgYfTBADcqgL4AaVTz6D04GqXHEpM2QC4YAM01g0mcDCUVNXUtHT1SA2NOc1UIECsIsBoAUUIIZGTpTA1CCXcoXE1kGAAyEs9vX39A9mCAejqYAAEoCABaZFkAB1QoDtxcEVVODShtXX0jUwtOfAGQAHcAMUqoP0gAQQ0NwgX8CQgAIRBvAXdAxgA%2BILUG5taO7t7%2BwdxhkLGwyaiYEzY-th3ZAQQiYaRtAQGfDwQjINpgTp9UEI%2B4SHoQYC4TBdPrA5FQOreeAnMACNoJQRsUCQWAsbi8fhCfiiBykpzSOS-JhwJC9LA4AgkciUah0SDGalxWEAOkIIBo8gpTJEYjZzjk0soPGUdwARErhCzHOrZLq2AbmarJByFMpGPaeSg0Py8EQyBQqLR6NKLSrWdaXIE7h4iOk2IDGgBJGCdHpocIGPQgTyEfA0TwicLFCD4HghfBxSAwAtZmBdDQANz8mgY4GQ0qp4GgrHp1gEtiS-vZLi5zAQTow2FdQo9ovoEqbIBlcoVSo79mNNs1ueQOsa%2BoZNjsXZNZrn28XgbtDv7fKHgvdIq9kB9m-bB7VNqDjRDaWQ4cltJgcQSYJSb8yNYcgkXtHTPAU3WFT0xSmSUp3rGd5B-Bk-1SdJAOyZBcmXbUYD1XUYAAHjaNoYAAORAHBwnwWBJhgRNIQ8Dw8AoWAPEGLhS3LZAqxOWsEQAQjYZD9CSNCMiyYCLhPXlnXPSDR2vCBpRExJ-3QySsIkZ9PFDd9GxpFs5kGZZVnWCAtmQHY9gOY5TlA085IgkcrxgidIHg2V5XkYzFhWHw1ibSzrP2I4SQEHDVzw9dfNMgLzOC3ZQrs0kzVi-yqk2bYkts8LpL7WTB2cy9oO9dKzKC7KbLC04dIKIo2HK%2BLKqsnKatJfKwKc4cSrHItgz0oA&eslintrc=N4KABGBEBOCuA2BTAzpAXGUEKQAIBcBPABxQGNoBLY-AWhXkoDt8B6WJgIwHsOATWgFtE%2BABbc%2B6KImjRu0SODABfEMqA&tsconfig=N4KABGBEDGD2C2AHAlgGwKYCcDyiAuysAdgM6QBcYoEEkJemy0eFYDAruuGAL4g9A&tokens=false)):

```typescript
const objectLiteralShenanigans = {
  methodShorthandSyntax() {
    return this;
  },
  methodLonghandSyntax: function () {
    return this;
  },
  somethingElseEntirely: true && function () {
    // @ts-expect-error
    return this;
  },
  arrowFunctionsAreAlwaysBound: () => {
    // @ts-expect-error
    return this;
  }
}

// eslint-disable-next-line @typescript-eslint/unbound-method
const { methodShorthandSyntax } = objectLiteralShenanigans;
console.log(methodShorthandSyntax.name) // "methodShorthandSyntax"
methodShorthandSyntax() === objectLiteralShenanigans.methodShorthandSyntax() // false

// I expect this to flag for the same reasons as the previous one.
const { methodLonghandSyntax } = objectLiteralShenanigans;
console.log(methodLonghandSyntax.name) // "methodLonghandSyntax"
methodLonghandSyntax() === objectLiteralShenanigans.methodLonghandSyntax() // false

const { somethingElseEntirely } = objectLiteralShenanigans;
console.log(somethingElseEntirely.name) // "" <-- Note that this is different from the previous one!
somethingElseEntirely() === objectLiteralShenanigans.somethingElseEntirely() // false

const { arrowFunctionsAreAlwaysBound } = objectLiteralShenanigans;
console.log(arrowFunctionsAreAlwaysBound.name) // "arrowFunctionsAreAlwaysBound"
arrowFunctionsAreAlwaysBound() === objectLiteralShenanigans.arrowFunctionsAreAlwaysBound() // true
arrowFunctionsAreAlwaysBound() === objectLiteralShenanigans // false
```

**Comment by bradzacher** [comment]
When inspecting the symbol from the member expression its `valueDeclaration` is `SyntaxKind.FunctionExpression`. The rule doesn't have handling for this because it assumes that the value declaration would be a property assignment.

Looking at the data - the parent of the function expression symbol is the property assignment.

So the fix here is to check and report the case of "if value declaration is function and parent is property assignment then report"
Should probably also handle the case of a `tbis: void` arg in the function expression.

**Comment by kirkwaiblinger** [comment]
tbh this one is almost harder to articulate in words than it is to fix 😆
